ruby/ql/lib/codeql/ruby/regexp/internal/RegExpTracking.qll

```diff
@@ -91,69 +55,172 @@ DataFlow::Node stringSink() {
+      reached(nodeFrom, t) and
+      reached(nodeTo, result) and
+      result = t.append(summary)
+    )
+    or
+    Input::additionalStep(nodeFrom, nodeTo) and
+    reached(nodeFrom, pragma[only_bind_into](t)) and
+    reached(nodeTo, pragma[only_bind_into](t)) and
+    result = t.continue()
+  }
+}
+
+pragma[nomagic]
+private predicate regFromString(DataFlow::LocalSourceNode n, DataFlow::CallNode call) {
+  exists(DataFlow::Node mid |
+    n.flowsTo(mid) and
+    call = API::getTopLevelMember("Regexp").getAMethodCall(["compile", "new"]) and
+    mid = call.getArgument(0)
+  )
+}
+
+private module StringReachInput implements ReachInputSig {
+  DataFlow::LocalSourceNode start(TypeTracker t) { result = strStart() and t.start() }
+
+  DataFlow::Node end() {
+    result = stringSink() or
+    regFromString(result, _)
+  }
+
+  predicate additionalStep(DataFlow::LocalSourceNode nodeFrom, DataFlow::LocalSourceNode nodeTo) {
+    exists(DataFlow::Node mid | nodeFrom.flowsTo(mid) |
+      // include taint flow through `String` summaries
+      TaintTrackingPrivate::summaryThroughStepTaint(mid, nodeTo, any(String::SummarizedCallable c))
+      or
+      // string concatenations, and
+      exists(CfgNodes::ExprNodes::OperationCfgNode op |
+        op = nodeTo.asExpr() and
+        op.getAnOperand() = mid.asExpr() and
+        op.getExpr().(Ast::BinaryOperation).getOperator() = "+"
+      )
+      or
+      // string interpolations
+      mid.asExpr() = nodeTo.asExpr().(CfgNodes::ExprNodes::StringlikeLiteralCfgNode).getAComponent()
+    )
+  }
 }
 
+private module StringReach = Reach<StringReachInput>;
+
 /**
  * Gets a node that has been tracked from the string constant `start` to some node.
  * This is used to figure out where `start` is evaluated as a regular expression against an input string,
  * or where `start` is compiled into a regular expression.
  */
 private DataFlow::LocalSourceNode trackStrings(DataFlow::Node start, TypeTracker t) {
-  result = backwards(_) and
-  (
-    t.start() and
-    start = result and
-    result = strStart()
-    or
-    exists(TypeTracker t2 | result = trackStrings(start, t2).track(t2, t))
+  t.start() and
+  start = result and
+  result = strStart() and
+  StringReach::reached(result, t)
+  or
+  exists(TypeTracker t2 | t = StringReach::stepReached(t2, trackStrings(start, t2), result))
+}
+
+pragma[nomagic]
+private predicate regFromStringStart(DataFlow::Node start, TypeTracker t, DataFlow::CallNode nodeTo) {
+  regFromString(trackStrings(start, t), nodeTo) and
+  exists(t.continue())
+}
+
+private module RegReachInput implements ReachInputSig {
+  DataFlow::LocalSourceNode start(TypeTracker t) {
+    result = regStart() and
+    t.start()
     or
-    // an additional step from string to string
-    exists(TypeTracker t2 | t2 = t.continue() |
-      step(trackStrings(start, t2).getALocalUse(), result, "string", "string")
-    )
-  )
+    regFromStringStart(_, t, result)
```

Comment:
But I see now, that it only makes sense to start flow when data is not stored inside content, so I'll fix that.

Reply:
Hmmm. That might have been a mistake from me then.

Reply:
Reusing the type tracker means we avoid spurious flow-through, where we might otherwise have a string that flows in at one call site, and out as a regular expression at another call site.

Note that we do check that the existing type tracker does not track data inside of contents.